File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public boolean helper(int arr[],int mid,int h){
        long count=0;
        for(int i:arr){
             count+= Math.ceil((double)i/mid);
        }
        return count<=h;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int low=1;
        int high=0;
        int max=0;
        for(int i:piles){
            max=Math.max(i,max);
        }
         high=max;
        int ans=high;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(helper(piles,mid,h)){
                ans=mid;
                high=mid-1;
            }
            else low=mid+1;
        }
        return ans;

        
    }
}

File: 907-koko-eating-bananas/koko-eating-bananas.py
import math 
class Solution:
    def minEatingSpeed(self, arr: list[int], h: int) -> int:
        def helper(mid , arr ,h):
            count =0
            for i in arr:
                count += math.ceil((i/mid))
            if count<=h:
                return True
            else :
                return False

        n = len(arr)
        low = 1
        high = max(arr)

        while(low<=high):
            mid=(low+high)//2
            if(helper(mid,arr,h)):
                ans = mid
                high=mid-1
            else :
                low = mid+1
        return ans
        
        